cache/cache.go

```diff
@@ -13,3 +13,268 @@
+				return
+			}
+		}
+	}()
+	return responseChan
+}
+
+// Ready reports whether the cache has finished its initial load.
+func (c *Cache) Ready() bool {
+	select {
+	case <-c.ready:
+		return true
+	default:
+		return false
+	}
+}
+
+// WaitReady blocks until the cache is ready or the ctx is cancelled.
+func (c *Cache) WaitReady(ctx context.Context) error {
+	select {
+	case <-c.ready:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Close cancels the private context and blocks until all goroutines return.
+func (c *Cache) Close() {
+	c.stop()
+	c.waitGroup.Wait()
+}
+
+// newWatchEventStream builds a single internal Watch event stream for
+// both historical replay (>= startRev) and live updates filtered by key.
+func (c *Cache) newWatchEventStream(ctx context.Context, key string, startRev int64) (<-chan *clientv3.Event, error) {
+	pred := func(k []byte) bool { return strings.HasPrefix(string(k), key) }
+
+	// TODO: re-enable arbitrary startRev support once we guarantee gap-free replay.
+	if startRev != 0 {
+		return nil, ErrUnsupportedStartRev
+	}
+
+	w := newWatcher(c.cfg.PerWatcherBufferSize, pred)
+	c.demux.add(w, 0)
+
+	go func() {
+		<-ctx.Done()
+		c.demux.remove(w)
+	}()
+
+	return w.eventQueue, nil
+}
+
+func serveWatchEvents(ctx context.Context, watchCtx *watchCtx, bufSize int) {
+	enqueueCh := make(chan *clientv3.Event, bufSize)
+
+	go func() {
+		for event := range enqueueCh {
+			watchCtx.broadcast(event)
+		}
+	}()
+
+	go func() {
+		defer close(enqueueCh)
+		watchRetryLoop(ctx, watchCtx, enqueueCh)
+	}()
+}
+
+func watchRetryLoop(ctx context.Context, watchCtx *watchCtx, enqueueCh chan<- *clientv3.Event) {
+	backoff := watchCtx.backoffStart
+	for {
+		opts := []clientv3.OpOption{
+			clientv3.WithPrefix(),
+			clientv3.WithProgressNotify(),
+			clientv3.WithCreatedNotify(), //robustness+
+		}
+		if oldestEvent := watchCtx.cache.history.PeekOldest(); oldestEvent != nil {
+			opts = append(opts,
+				clientv3.WithRev(oldestEvent.Kv.ModRevision+1))
+		}
+		watchCh := watchCtx.cache.watcher.Watch(ctx, watchCtx.cache.prefix, opts...)
+
+		if err := readWatchChannel(ctx, watchCh, watchCtx.cache, watchCtx.cache.history, watchCtx.cache.demux, enqueueCh, watchCtx.onFirstResponse); err == nil {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(backoff):
+		}
+		if backoff < watchCtx.backoffMax {
+			backoff *= 2
+		}
+	}
+}
+
+// readWatchChannel reads an etcd Watch stream into History and enqueueCh, returning nil on cancel or first error.
+func readWatchChannel(
+	ctx context.Context,
+	watchChan clientv3.WatchChan,
+	cache *Cache,
+	ring *ringBuffer,
+	demux *demux,
+	enqueueCh chan<- *clientv3.Event,
+	onFirstResponse func(),
+) error {
+	for resp := range watchChan {
+		onFirstResponse()
+
+		if err := resp.Err(); err != nil {
+			if errors.Is(err, rpctypes.ErrCompacted) {
+				ring.RebaseHistory()
+				select {
+				case <-cache.ready:
+					cache.ready = make(chan struct{})
```

Comment:
Assigning a channel will be racy, but fix is complicated. Can you leave a TODO to fix this?